Fix issue 1185

dabiged opened the pull request:
I have made some minor changes to cfd example notebooks 1 - 4 to address bug #1185 .

Also I noticed that there was a mistake in the numpy implementation in these notebooks that was not noticeable until I changed the domain from nxn to nx2n. 

This is my first time contributing to a large project and I am not sure what I am doing. 
Guidance welcome.

rhodrin commented:
Many thanks for the contribution. We'll take a look over the changes and reply in more detail soon.

mloubout commented:
Thanks a lot for the PR this is great, will look at it. The notebook should be run from top to bottom then saved so that the figures and cell output are present in it.

dabiged commented:
Should I resubmit a pull request with the notebook in the appropriate format?

mloubout commented:
No leave it open, just rerun the notebooks and push it as a new commit.

dabiged commented:
I have rerun all the notebooks and pushed the new commit.

mloubout commented:
Great looks really nice.

mloubout commented:
Will go through it over the weekend/mondat. Does it fix all the points in the linked issue? Just to know if we should close it once this is merged.

dabiged commented:
As mentioned above this pull request does not fulfill all the requirements to close #1185. 
Notably: No changes are made to the 05 or 06 notebooks, and some enhancements have not yet been incorporated

I think the 05 notebook has a real issue in that the x,y convention changes to y,x, to align with previous work (see developer note in preamble). This is very confusing especially around the boundary conditions (as noted by @jkwashbourne in #1185) and personally I think interrupts the flow of the tutorial series. To me the best thing to do would be to rewrite this notebook without this change in convention. I am happy to give this a go.

dabiged commented:
Below is a check list from #1185 as to what is fixed in the current commit for notebooks 1-4 .

**01_convection.ipynb**

- [ ] Suggest noting that u[:,-1] notation is shorthand for the maximum
- [ ] In cell [4] last sentence, it was not clear to me that the 3d array u[t,x,y] could be indexed as u[0] to provide u[0,x,y]. Is this a numpy shorthand?
- [ ] Cell [6]: wondering why the "1.0*" doesn't disappear?
- [ ] Cell [6]: you refer to "s" and "h" but they are not seen here. Did the output from print(eq) change? We do see "h_x" and "h_y" in cell [7].
- [x] Cell [7]: discussion uses "h" but stencil uses "h_x" and "h_y". Note there is continuing confusion about "s", "dt".
- [ ] I needed "%matplotlib inline" here also, must be my setup is whack
- [ ]     Cell [9]: I was expecting a smaller tolerance after implementing the BCs, what gives? Should the Devito result with no optimization be float eps equivalent to the numpy one?

**01_convection_revisited.ipynb**

- [x] s/exampe/example/
- [ ] Cell [9] uses "nx - 1" instead of "-1" notation, consistency?
- [ ] Again schizo on "-1" and "n-1" notation for end of array

**02_convection_nonlinear.ipynb**

- [x] Cell [1] here has "%matplotlib inline", suggest adding to the others in the first cell
- [x] "s/hte/the/"
- [x] "s/woth/with/"
- [ ] Cell [8]: A couple more words on the "automated staggering of vectorial function" would be super useful

**03_diffusion.ipynb**

- [ ] Cell [3]: omg such a python n00b, I had to go look up the "//" operator. Is it worth noting what that operator does? Maybe not since you have the plot just below.
- [x] Cell [4] is a partial duplicate of cell [3]?
- [ ] Cell [7], [8] there is inconsistency in line feeds ... [not sure anyone should care about this level of pedantry]
- [ ] Cell [7] and [8] seem redundant

**03_diffusion_nonuniform.ipynb**

- [x] Suggest clarification: "This example is an enhancement of 03_diffusion in terms of having non-uniform viscosity as opposed to 03_diffusion" --> "This example is an enhancement of 03_diffusion in terms of having non-uniform viscosity"
- [ ] Suggest clarification: "In our case ν(x,y) is depended on the material properties as we may have different viscosity for a (x,y) pair. So ν(x,y) is not uniform." --> "In our case v is not uniform and ν(x,y) represents spatially variable viscosity."
- [ ] Cell [4] and [5]: suggest not re-using all the same language from 03_diffusion.ipynb about the "trick" of the laplace operator, as you already did that.
- [ ]  Cell [7] and [8] seem redundant

**04_burgers.ipynb**

- [x] Cell [1]: "# Increase number of batches to extend evolution in time" is used way too many times

dabiged commented:
Updated the checklist below. My comments on the commit are in _italics_

**01_convection.ipynb**

- [x] Suggest noting that u[:,-1] notation is shorthand for the maximum
- [x] In cell [4] last sentence, it was not clear to me that the 3d array u[t,x,y] could be indexed as u[0] to provide u[0,x,y]. Is this a numpy shorthand?
- [x] Cell [6]: wondering why the "1.0*" doesn't disappear?
- [x] Cell [6]: you refer to "s" and "h" but they are not seen here. Did the output from print(eq) change? We do see "h_x" and "h_y" in cell [7].
- [x] Cell [7]: discussion uses "h" but stencil uses "h_x" and "h_y". Note there is continuing confusion about "s", "dt".
- [x] I needed "%matplotlib inline" here also, must be my setup is whack. 
     _"%matplotlib inline" is in cell 1 now._
- [ ]     Cell [9]: I was expecting a smaller tolerance after implementing the BCs, what gives? Should the Devito result with no optimization be float eps equivalent to the numpy one? 
    _I cannot explain the difference between the numpy implementation and the devito implementation. The difference is up to +/- 0.2_

Other notes:
Fixed bug in Numpy code.
Added text to relationship between numpy, analytical and devito implementation prior to cell 8.

**01_convection_revisited.ipynb**

- [x] s/exampe/example/
- [x] Cell [9] uses "nx - 1" instead of "-1" notation, consistency? 
     _Added explanation as to why this occurs_
- [x] Again schizo on "-1" and "n-1" notation for end of array. 
     _As above_

**02_convection_nonlinear.ipynb**

- [x] Cell [1] here has "%matplotlib inline", suggest adding to the others in the first cell
- [x] "s/hte/the/"
- [x] "s/woth/with/"
- [ ] Cell [8]: A couple more words on the "automated staggering of vectorial function" would be super useful. 
     _Not sure I can offer much here_

**03_diffusion.ipynb**

- [x] Cell [3]: omg such a python n00b, I had to go look up the "//" operator. Is it worth noting what that operator does? Maybe not since you have the plot just below.
- [x] Cell [4] is a partial duplicate of cell [3]?
- [x] Cell [7], [8] there is inconsistency in line feeds ... [not sure anyone should care about this level of pedantry]
- [ ] Cell [7] and [8] seem redundant. 
     _I disagree. The second shows how easy it is to re-initialise and rerun once everything is setup._

Other notes:
Added nsimplify to equation displays in cells 4 and 5. This replaces -1.0*x with -x etc.

**03_diffusion_nonuniform.ipynb**

- [x] Suggest clarification: "This example is an enhancement of 03_diffusion in terms of having non-uniform viscosity as opposed to 03_diffusion" --> "This example is an enhancement of 03_diffusion in terms of having non-uniform viscosity"
- [x] Suggest clarification: "In our case ν(x,y) is depended on the material properties as we may have different viscosity for a (x,y) pair. So ν(x,y) is not uniform." --> "In our case v is not uniform and ν(x,y) represents spatially variable viscosity."
- [x] Cell [4] and [5]: suggest not re-using all the same language from 03_diffusion.ipynb about the "trick" of the laplace operator, as you already did that. 
    _Delete cell 5 and make comment prior to cell 4 around the use of the laplacian_
- [x]  Cell [7] and [8] seem redundant. 
    _Deleted Cell 8_

Other notes:
Added nsimplify to equation displays in cell 5.

**04_burgers.ipynb**

- [x] Cell [1]: "# Increase number of batches to extend evolution in time" is used way too many times

Other notes:
Removed eq_u declaration at the end of cell 6. This was never used. It also used the identity: div(grad(U)) = laplacian(U) for no reason.
Made a copy of the analytical equations above cell 8 for clarity. This allows the reader to see why the stencil is trivial.
Merge cells 8 and 9.
Removed cell 10 (div(grad(U))) was all it contained.

dabiged commented:
I have just pushed the changes to the 05 and 06 notebooks. Summary of changes and how they relate to #1185 below.

**05_laplace.ipynb**
- [x] The most confusing thing about this tuto is the BCs. I think a little more time thinking about how to make it more easy to understand would be super useful.
   _Added an explanation below cell 1 around how we will implement the BCs_
- [x] Cell [3]: "s/Out initial condition/Our initial condition/"
- [x] Cell [3]: are the labels wrong on the plot due to the (y,x) weirdness? The BC "# p = y @ x = 2" looks more like "# p = x/2 @ y = 1"
- [x] Same confusion in cell [7]
- [x] Cell [6]: "s/where we wet/where we want"
- [ ] I think cells [6] and [7] need more words
- [x] I note that the first figure from cell [3] and the first figure from cell [8] look different. This appears to be some kind of rat dropping at x=0, y=1 
    _This is due to the Neuman boundary condition. The 'rat dropping' is copying the second to last value into the last line which distorts the straight line. There is nothing to fix here._

Other notes:
Added timing magic command on the last 2 cells to show that the memory copy IS slow.
Removed the y,x nomenclature from this notebook and replaced it with the x,y paradigm. This is inline with the rest of the notebooks. **This is a BIG change**. The images in the notebook now match those in the original CFDPython tutorials.

**06_poisson.ipynb**
- [x] above cell [1]: "−100 at i=nx∗3/4,j=3/4ny" --> "−100 at i=3/4 nx,j=3/4 ny"
- [x] Cell [3]: the comment "# Now with Devito we will turn p into TimeFunction object to make all the buffer switching implicit" seems to jumping the gun, and belongs in a cell below
    _These comments deleted_
- [x] Cell [3]: More schizophrenia with notation [nx-1,:] or [-1,:]     
    _Actually cell 6. I addressed this in a prior notebook and it is due to some Boundary conditions being set on numpy array objects which allow [-1] addressing, and some being set on sympy functions which do not._
- [x] Suggest adding a comparison of the final pure python code and the devito code for a large enough model to demonstrate a big performance delta. 
    _Added %%time magic commands to relevant cells. Speedup is now obvious_


Other notes:
Fixed typo souce -> source in preamble.
Merged blocks of 6 sequential cells into 3; 1 to setup the problem, 1 to run it and measure runtime, and 1 to show the result graphically.
Added some text on speedups prior to the last section.

FabioLuporini commented:
@dabiged , many thanks for this. On a first glance it really, really looks like the cleanup we needed!

A few things:

* There now are conflicts with the notebooks in master. Not sure if they've been updated recently or you were working from an older devito, but we need to fix them. There are a few tools to diff notebooks, such as [this one](https://github.com/jupyter/nbdime) (just found out). Talk to us (slack is also fine) if you can't get your head around these conflicts

* I've now formally assigned the review of this PR to @georgebisbas , @rhodrin , and @mloubout 

* I'm totally fine, obviously, if this PR doesn't fix the whole of 1185 -- it definitely goes in the right direction.

dabiged commented:
I have rebased this branch on upstream and resolved the conflicts (all of which were in the OUTPUT cells of the notebooks).

georgebisbas commented:
Progress looks great @dabiged ! I believe its worth that we help some more to finish up the issue. 
P.s. You can edit the initial comment: https://github.com/devitocodes/devito/pull/1360#issue-436440310
and update it with the changes list using the tick boxes instead of pasting more than once the checklist ;-) .

dabiged commented:
> P.s. You can edit the initial comment: [#1360 (comment)](https://github.com/devitocodes/devito/pull/1360#issue-436440310)
> and update it with the changes list using the tick boxes instead of pasting more than once the checklist ;-) .

What magic is this?!?!?

rhodrin commented on examples/cfd/01_convection.ipynb:
```diff
@@ -120,14 +120,15 @@
+    "                              (c * dt / dx * (un[1:, 1:] - un[:-1, 1:])))\n",
+    "\n",
+    "    # Apply boundary conditions. \n",
+    "    # Note: -1 here is the last index in the array, not the one at x=-1 or y=-1.\n",
```
I'd move this note to below the expressions and modify it slightly to something along the lines of:
"Note that in the above expressions the `NumPy` index `-1` corresponds to the final point of the array along the indexed dimension, i.e. here `u[-1, :]` is equivalent to `u[80, :]`."

rhodrin commented on examples/cfd/01_convection.ipynb:
```diff
@@ -206,8 +209,8 @@
     "u = TimeFunction(name='u', grid=grid)\n",
-    "init_hat(field=u.data[0], dx=dx, dy=dy, value=2.)\n",
-    "\n",
+    "init_hat(field=u.data[0,:,:], dx=dx, dy=dy, value=2.)\n",
```
Since the note has been added why not keep `u.data[0]` here?

rhodrin commented on examples/cfd/01_convection.ipynb:
```diff
@@ -261,23 +264,25 @@
    "metadata": {},
    "source": [
-    "The careful reader will note that this is now the equivalent symbolic expression to the RHS term of the NumPy code we showed earlier - only with `dx` and `dy` denoted as `h` and `dt` denoted as `s`, while `u(t, x, y)`, `u(t, x - h, y)` and `u(t, x, y - h)` denote the equivalent of $u^{n}_{i, j}$, $u^{n}_{i-1, j}$ and $u^{n}_{i, j-1}$ respectively. \n",
+    "The careful reader will note that this is now the equivalent symbolic expression to the RHS term of the NumPy code we showed earlier - only with `dx` and `dy` denoted as `h_x` and `h_y` respectively, while `u(t, x, y)`, `u(t, x - h, y)` and `u(t, x, y - h)` denote the equivalent of $u^{n}_{i, j}$, $u^{n}_{i-1, j}$ and $u^{n}_{i, j-1}$ analytically and `u[1:, 1:]`, `u[:-1, 1:]` and `u[1:, :-1 ...
```
'...this is now the equivalent symbolic expression to the RHS term of the...' -> '...this is equivalent to the symbolic expression for the RHS of the..."

rhodrin commented on examples/cfd/02_convection_nonlinear.ipynb:
```diff
@@ -338,8 +305,8 @@
-    "We now have a vectorial equation. Unlike hte previous case, we do not need to play woth legt/right derivatives\n",
-    "as the automated staggering of vectorial function takes care of it."
+    "We now have a vectorial equation. Unlike the previous case, we do not need to play with left/right derivatives\n",
+    "as the automated staggering of the vectorial function takes care of it."
```
Unlike **in** the previous case...
...takes care of this.

rhodrin commented:
I think once all the above comments have been addresses, the branch has been rebased on top of master and conflicts resolved, and then all the notebooks re-ran, when green, should be gtg.

mloubout commented:
I think looks really nice, thanks for all of that, should be gtg with the minor comments fixed once CI is green (just rerunning it top to bottom should fix CI).

dabiged commented:
I will take a look at this today.

dabiged commented on examples/cfd/06_poisson.ipynb:
```diff
@@ -278,17 +281,17 @@
-      "CPU times: user 1 µs, sys: 1e+03 ns, total: 2 µs\n",
-      "Wall time: 4.05 µs\n"
+      "CPU times: user 169 ms, sys: 9.68 ms, total: 178 ms\n",
+      "Wall time: 179 ms\n"
```
I changed the magic command for this cell from %time (time this command) to %%time (time this cell).
The pseudo time operator is now twice as slow as the buffer stepping scheme.

dabiged replied:
This is now fixed. As pointed out by @FabioLuporini the last cell included the compilation of the operator as well as it running. I have moved the time command to only run on the operator now.

FabioLuporini commented:
Awesome, thanks a lot @dabiged . Merging...
